meetings/working-groups/collection-literals/Core-interface-target-type-proposal.md

```diff
@@ -0,0 +1,247 @@
+
+For 'brevity/universality', this still strongly satisfies our goals.  Users will be able to use the succinct collection-expression form for all these APIs with ease. However, compared to option 2, we now see *large* wins in both safety and performance. 
+
+Starting with safety, we now have a good option for everyone exposing data through any of those three interfaces.  Data is safe by default (a good choice for all users).  Analyzers do not trigger, preventing our own tooling from pushing people away from using our cohesive story here.  Users who do not care about safety continue not to care, while users for whom it is very important are extremely satisfied that they can trust this feature and that it lives up to the stated goal that it will make the smart choices they can depend on, and they do not have to ban this feature, or recommend others steer away from it.  This approach also naturally extends out to when we do dictionary-literals.  There, it will also be the case that for `IReadOnlyDictionary<TKey, Value>` we will not want to expose a mutable type (like `Dictionary<TKey, TValue>`) for safety reasons.
+
+While the wins now with 'safety' are definitely welcome, the largest benefits come in the "performance" category.  Specifically, not having to name a particular type for the read-only interfaces opens up many areas of optimization that would simply be unavailable when having to pick an existing generalized type. This would also live up to the promise that we will do an excellent job with perf, and that it would be *very* hard for a user to do the same, and practically impossible for them to do so with simple syntax.  Specifically, all of the following are available as things we could do to heavily optimize here when target-typing `IEnumerable<T>/IReadOnlyCollection<T>/IReadOnlyList<T>:
+
+1. Empty collection expressions can always be collapsed down to a single empty singleton (per `T` element type) across the entire program.
+
+1. We can use features like inline-arrays to generate instances that inline all their elements are literally no larger than space for the elements themselves.  A count would not even need to be stored as it could just be hard-coded as the value returned by `int Count { get; }`.  While this might generate many types in the assembly (one type per unique count), it's worth noting that this is what we *already* do for arrays *today*, and what we will be doing for the inline-array types we generate for spans/params-spans.
+
+1. We can borrow a trick from `yield iterators` where the iterator returned from `IEnumerator<T> GetEnumerator()` literally *reuses* the `this` instance to be its own iterator *if* it has never been iterated yet, and the calling thread is the same one that generated the `IEnumerable<T>`.  This means that if you have:
+
+    ```c#
+    DoSomething([1, 2, 3]);
+    ```
+
+    Then this may be literally a *single* allocation, *including* for when it is likely almost immediately iterated.
```

Comment:
I'm not entirely sure why the Markdown syntax highlighting broke so badly from this point forward (not ending the code syntax highlighting) but maybe it's due to not having the expected number of spaces of indentation?

````suggestion
   ```c#
   DoSomething([1, 2, 3]);
   ```

   Then this may be literally a *single* allocation, *including* for when it is likely almost immediately iterated.
````